### PR TITLE
Implement the runtime distributions for all the existing probability distribution except GPs

### PR DESCRIPTION
**Pull request by zhenwendai**
*Description of changes:*

It contains the implementation of the runtime distributions for all the existing probability distribution except GPs. The log_pdf and draw_samples implemention in the original distribution files are removed.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: mxfusion/runtime/distributions/multivariate_normal.py [hunk 1]
```diff
@@ -124,28 +117,116 @@ def kl_divergence(self, other):
...
+        """
+        :param mean: Mean of the normal distribution. Shape: (n_samples, n_outputs, n_dim)
+        :type mean: MXNet NDArray
+        :param covariance: Covariance matrix of the distribution. Shape: (n_samples, n_outputs, n_dim, n_dim)
+        :type covariance: MXNet NDArray
+        """
+        super(MultivariateNormalMeanPrecisionRuntime, self).__init__()
+
+        if precision is None or mean.shape != precision.shape[:-1] or precision.shape[-2]!=precision.shape[-1]:
+            raise InferenceError('Mean and covariance shapes inconsistent. Mean shape: {}. Covariance shape: {}'.format(
```
Comment (meissnereric) [inline mxfusion/runtime/distributions/multivariate_normal.py]: covariance -> precision

File: mxfusion/runtime/distributions/multivariate_normal.py [hunk 2]
```diff
@@ -124,28 +117,116 @@ def kl_divergence(self, other):
...
+                mean.shape, precision.shape))
+
+        if precision is None and cholesky_factor is None:
+            raise InferenceError('Either covariance or cholesky_factor needs to be specified.')
```
Comment (meissnereric) [inline mxfusion/runtime/distributions/multivariate_normal.py]: covariance -> precision

**Review by meissnereric — APPROVED** [review]
Looks good to me, just a few typo comments.

**Comment by meissnereric** [comment]
General question for runtime distributions:

https://github.com/zhenwendai/MXFusion/blob/new_dist_impl/mxfusion/components/distributions/distribution.py#L84

Why do we create a new distribution on the fly here instead of just passing variables (or doing the kwarg parsing from lines 55/56 first) to runtime_distribution.log_pdf() and using it as a static function kind of thing? This would save us creating a new object that we immediately throw away every time we call log_pdf/draw_samples on a factor yeah?

File: mxfusion/runtime/distributions/multivariate_normal.py (hunk 1 above)
Comment (zhenwendai) [reply mxfusion/runtime/distributions/multivariate_normal.py]: Done.

File: mxfusion/runtime/distributions/multivariate_normal.py (hunk 2 above)
Comment (zhenwendai) [reply mxfusion/runtime/distributions/multivariate_normal.py]: Done.

**Comment by zhenwendai** [comment]
> General question for runtime distributions:
> 
> https://github.com/zhenwendai/MXFusion/blob/new_dist_impl/mxfusion/components/distributions/distribution.py#L84
> 
> Why do we create a new distribution on the fly here instead of just passing variables (or doing the kwarg parsing from lines 55/56 first) to runtime_distribution.log_pdf() and using it as a static function kind of thing? This would save us creating a new object that we immediately throw away every time we call log_pdf/draw_samples on a factor yeah?

The ```log_pdf``` function in the ```Distribution``` class is only to keep compatibility with the existing codes, but the introduction of the runtime distributions is to support the creation of distributions that are not defined in the model/posterior definition.

**Pull request merged**